### ISSUE TITLE
[rt.openssl.org #4156] Wrong expectation of value of _IOB_ENTRIES with Visual Studio 2015

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#2404]
Migrated from [rt.openssl.org#4156](https://rt.openssl.org/Ticket/Display.html?id=4156) (status was 'open')

Requestors:
* hannes.mezger@ascolab.com

From hannes.mezger@ascolab.com on 2015-11-25 19:33:00:
```
Hi,

when using OpenSSL (1.0.1p) in our applications, we get the error 'No
OPENSSL_Applink' when calling PEM_read_X509_CRL. We build both OpenSSL
and the application using the same C runtime configuration (/MDd), using
Visual Studio 2015 on Windows 10.

After some hours of debugging and comparing with the same scenario using
Visual Studio 2010, I discovered that _IOB_ENTRIES is defined
differently in the two versions of Visual Studio:

VS2015: c:\Program Files (x86)\Windows
Kits\10\Include\10.0.10150.0\ucrt\stdio.h: #define _IOB_ENTRIES 3
VS2010: c:\Program Files (x86)\Microsoft Visual Studio
10.0\VC\include\stdio.h: #define _IOB_ENTRIES 20

Now there's code in openssl-1.0.1p\crypto\bio\bss_file.c that seems to
expect _IOB_ENTRIES to have the value 20:

#   if defined(__MINGW32__) && defined(__MSVCRT__) && !defined(_IOB_ENTRIES)
#    define _IOB_ENTRIES 20
#   endif
#   if defined(_IOB_ENTRIES)
        /* Safety net to catch purely internal BIO_set_fp calls */
        if ((size_t)ptr >= (size_t)stdin &&
            (size_t)ptr < (size_t)(stdin + _IOB_ENTRIES))
            BIO_clear_flags(b, BIO_FLAGS_UPLINK);
#   endif
#  endif
#  ifdef UP_fsetmod
        if (b->flags & BIO_FLAGS_UPLINK)
            UP_fsetmod(b->ptr, (char)((num & BIO_FP_TEXT) ? 't' : 'b'));
        else
#  endif

The 'safety net' does not work here, as _IOB_ENTRIES is too small in
VS2015 and the flags don't get cleared (as for VS2010), so UP_fsetmod is
being called which leads to the 'No OPENSSL_Applink' error.

We kindly ask you to provide a solution for our problem, as we currently
are not able to use OpenSSL in combination with VS2015. Linking the
applink.c into our applications is no option for us, as we would have to
deploy multiple C runtimes (one for OpenSSL and one for the application).

Best regards,

Hannes Mezger

-- 
*Hannes Mezger*
ascolab GmbH
Tel.: +49 9131 691 124
Fax: +49 9131 691 128
Web: http://www.ascolab.com
GPG-KeyId: 46EC1280
GPG-Fingerprint: D28A DE55 05FA 2FF3 8628 323A 1D6E 9EC8 46EC 1280

--
ascolab GmbH - Am Weichselgarten 7 - 91058 Erlangen - Germany
Handelsregister/Commercial Register: Amtsgericht FÃ¼rth HRB 9360
GeschÃ¤ftsfÃ¼hrer/Managing Directors: Gerhard Gappmeier, Matthias Damm,
Uwe SteinkrauÃ
This email contains confidential information and is for the exclusive
use of the addressee. If you are not the addressee, then any
distribution, copying or use of this email is prohibited. If received in
error, please advise the sender and delete immediately. We accept no
liability for any loss or damage suffered by any person arising from use
of this email.

```

From appro@openssl.org on 2015-12-05 10:13:22:
```
Hi,

> when using OpenSSL (1.0.1p) in our applications, we get the error 'No
> OPENSSL_Applink' when calling PEM_read_X509_CRL. We build both OpenSSL
> and the application using the same C runtime configuration (/MDd), using
> Visual Studio 2015 on Windows 10.

Original intention with OPENSSL_Applink was to facilitate certain things
for old code. This kind of implies that new code should be written in
such manner that would *avoid* Applink. For this reason I would
recommend to modify application and switch to PEM_read_bio_X509_CRL. But
the keyword is that you should let OpenSSL open the file for you by
calling BIO_new_file. As opposite to opening yourself and calling
BIO_new_fp that is.

> After some hours of debugging and comparing with the same scenario using
> Visual Studio 2010, I discovered that _IOB_ENTRIES is defined
> differently in the two versions of Visual Studio:
> 
> VS2015: c:\Program Files (x86)\Windows
> Kits\10\Include\10.0.10150.0\ucrt\stdio.h: #define _IOB_ENTRIES 3
> VS2010: c:\Program Files (x86)\Microsoft Visual Studio
> 10.0\VC\include\stdio.h: #define _IOB_ENTRIES 20
> 
> Now there's code in openssl-1.0.1p\crypto\bio\bss_file.c that seems to
> expect _IOB_ENTRIES to have the value 20:
> 
> #   if defined(__MINGW32__) && defined(__MSVCRT__) && !defined(_IOB_ENTRIES)
> #    define _IOB_ENTRIES 20
> #   endif
> #   if defined(_IOB_ENTRIES)
>         /* Safety net to catch purely internal BIO_set_fp calls */
>         if ((size_t)ptr >= (size_t)stdin &&
>             (size_t)ptr < (size_t)(stdin + _IOB_ENTRIES))
>             BIO_clear_flags(b, BIO_FLAGS_UPLINK);
> #   endif
> #  endif
> #  ifdef UP_fsetmod
>         if (b->flags & BIO_FLAGS_UPLINK)
>             UP_fsetmod(b->ptr, (char)((num & BIO_FP_TEXT) ? 't' : 'b'));
>         else
> #  endif
> 
> The 'safety net' does not work here, as _IOB_ENTRIES is too small in
> VS2015 and the flags don't get cleared (as for VS2010), so UP_fsetmod is
> being called which leads to the 'No OPENSSL_Applink' error.

Well, commentary is probably misleading in the context, because "purely
internal BIO_set_fp" refers to BIO_set_fp called with FILE * created
internally, rather than just any BIO_set_fp called internally. So that
if application passes FILE * it created itself, it should expect UPLINK
calls. The fact that it worked for you without applink.c in VS2010 is
circumstantial.

> We kindly ask you to provide a solution for our problem, as we currently
> are not able to use OpenSSL in combination with VS2015. Linking the
> applink.c into our applications is no option for us, as we would have to
> deploy multiple C runtimes (one for OpenSSL and one for the application).

Last sentence is false in sense that linking applink.c into application
does *not* force you into using multiple C runtimes. It allows to use
multiple runtimes, but does not prohibit/prevent using single one. In
other words if this is the only reason for not linking the applink.c,
then it should be disregarded. But, once again, the right thing to do is
to adhere to PEM_read_bio_X509_CRL. This applies to all interfaces that
accept FILE *.

```

From hannes.mezger@ascolab.com on 2015-12-08 14:31:26:
```
Hi,

thanks for the detailed explanation. We can change our code to use the
BIO-functionality instead of using FILE pointers, so far this works fine
and even simplifies our code.

But two issues remain:

1. There is still no explanation on why it worked fine with Visual
Studio 2005 - 2013 (yes, checked all of those). To me it seems that the
code either doesn't handle the changed _IOB_ENTRIES correctly or is not
prepared to work with the new Windows 10 SDK.

2. The more important issue: we use 'ERR_print_errors_fp(stderr)' to
print OpenSSL errors directly to stderr. This doesn't work, too. But it
can't be fixed by using the 'ERR_print_errors()' function instead, as
creating the BIO using 'BIO_new_fp(stderr, BIO_NOCLOSE)' also fails with
the same error (no Applink). Is there a recommentation what to use in
that case?

Best regards,

Hannes

*Hannes Mezger*
ascolab GmbH
Tel.: +49 9131 691 124
Fax: +49 9131 691 128
Web: http://www.ascolab.com
GPG-KeyId: 46EC1280
GPG-Fingerprint: D28A DE55 05FA 2FF3 8628 323A 1D6E 9EC8 46EC 1280

--
ascolab GmbH - Am Weichselgarten 7 - 91058 Erlangen - Germany
Handelsregister/Commercial Register: Amtsgericht FÃ¼rth HRB 9360
GeschÃ¤ftsfÃ¼hrer/Managing Directors: Gerhard Gappmeier, Matthias Damm,
Uwe SteinkrauÃ
This email contains confidential information and is for the exclusive
use of the addressee. If you are not the addressee, then any
distribution, copying or use of this email is prohibited. If received in
error, please advise the sender and delete immediately. We accept no
liability for any loss or damage suffered by any person arising from use
of this email.

Am 05.12.2015 um 11:13 schrieb Andy Polyakov via RT:
> Hi,
>
>> when using OpenSSL (1.0.1p) in our applications, we get the error 'No
>> OPENSSL_Applink' when calling PEM_read_X509_CRL. We build both OpenSSL
>> and the application using the same C runtime configuration (/MDd), using
>> Visual Studio 2015 on Windows 10.
> Original intention with OPENSSL_Applink was to facilitate certain things
> for old code. This kind of implies that new code should be written in
> such manner that would *avoid* Applink. For this reason I would
> recommend to modify application and switch to PEM_read_bio_X509_CRL. But
> the keyword is that you should let OpenSSL open the file for you by
> calling BIO_new_file. As opposite to opening yourself and calling
> BIO_new_fp that is.
>
>> After some hours of debugging and comparing with the same scenario using
>> Visual Studio 2010, I discovered that _IOB_ENTRIES is defined
>> differently in the two versions of Visual Studio:
>>
>> VS2015: c:\Program Files (x86)\Windows
>> Kits\10\Include\10.0.10150.0\ucrt\stdio.h: #define _IOB_ENTRIES 3
>> VS2010: c:\Program Files (x86)\Microsoft Visual Studio
>> 10.0\VC\include\stdio.h: #define _IOB_ENTRIES 20
>>
>> Now there's code in openssl-1.0.1p\crypto\bio\bss_file.c that seems to
>> expect _IOB_ENTRIES to have the value 20:
>>
>> #   if defined(__MINGW32__) && defined(__MSVCRT__) && !defined(_IOB_ENTRIES)
>> #    define _IOB_ENTRIES 20
>> #   endif
>> #   if defined(_IOB_ENTRIES)
>>         /* Safety net to catch purely internal BIO_set_fp calls */
>>         if ((size_t)ptr >= (size_t)stdin &&
>>             (size_t)ptr < (size_t)(stdin + _IOB_ENTRIES))
>>             BIO_clear_flags(b, BIO_FLAGS_UPLINK);
>> #   endif
>> #  endif
>> #  ifdef UP_fsetmod
>>         if (b->flags & BIO_FLAGS_UPLINK)
>>             UP_fsetmod(b->ptr, (char)((num & BIO_FP_TEXT) ? 't' : 'b'));
>>         else
>> #  endif
>>
>> The 'safety net' does not work here, as _IOB_ENTRIES is too small in
>> VS2015 and the flags don't get cleared (as for VS2010), so UP_fsetmod is
>> being called which leads to the 'No OPENSSL_Applink' error.
> Well, commentary is probably misleading in the context, because "purely
> internal BIO_set_fp" refers to BIO_set_fp called with FILE * created
> internally, rather than just any BIO_set_fp called internally. So that
> if application passes FILE * it created itself, it should expect UPLINK
> calls. The fact that it worked for you without applink.c in VS2010 is
> circumstantial.
>
>> We kindly ask you to provide a solution for our problem, as we currently
>> are not able to use OpenSSL in combination with VS2015. Linking the
>> applink.c into our applications is no option for us, as we would have to
>> deploy multiple C runtimes (one for OpenSSL and one for the application).
> Last sentence is false in sense that linking applink.c into application
> does *not* force you into using multiple C runtimes. It allows to use
> multiple runtimes, but does not prohibit/prevent using single one. In
> other words if this is the only reason for not linking the applink.c,
> then it should be disregarded. But, once again, the right thing to do is
> to adhere to PEM_read_bio_X509_CRL. This applies to all interfaces that
> accept FILE *.
>
>

```

**Comment by jay** [comment]
Ref https://github.com/Microsoft/openssl/blob/OpenSSL_1_0_2_WinRT-stable/crypto/bio/bss_file.c#L323-L337

**Comment by bagder** [comment]
Without this issue fixed, a large number of OpenSSL-using applications will be broken when OpenSSL is built with VS2015.

**Comment by richsalz** [comment]
can we get a simple explanation of what needs to be fixed?  and for which branches?

**Comment by jay** [comment]
According to the comments in the link I referenced there are just the 3 IOB entries in the MS Universal CRT and they are no longer in an array and so the addresses must be retrieved using [__acrt_iob_func](https://msdn.microsoft.com/en-us/library/dn727675.aspx) instead of just doing stdin + n. I can't find documentation for it. It seems to me like it would be easier to do something like
if ptr == stdin || ptr == stdout || ptr == stderr
  clear flags

/cc @khouzam

**Comment by khouzam** [comment]
I made the change in the Microsoft fork to address this. 

https://github.com/Microsoft/openssl/commit/3e8a22#diff-b72a7d2b99757de10481d8ff50e29c16

Taking the change should be simple, but there will most likely need to adapt the #define specified to detect VS2015 vs other versions of Visual Studio.

I can provide more information and help validate if necessary,.

**Comment by bagder** [comment]
@khouzam awesome, can you provide a patch/pull request that would be what you suggest should be merged, including such improved compiler version checks?

**Comment by sunnyap** [comment]
@khouzam Not all OpenSSL Application uses Microsoft/OpenSSL. We need to get this fix in OpenSSL itself.

**Comment by richsalz** [comment]
we'll take this into openssl if someone makes a PR.

**Comment by levitte** [comment]
Couldn't we simply backport 800b299b0a53c871be57c4b2c97c95d6642489ea ?

**Comment by levitte** [comment]
See #2701

**Comment by sunnyap** [comment]
@levitte #2701 didn't worked for me.

**Comment by levitte** [comment]
Interesting.  Have you had a chance to try out 1.1.0?  Does that one work for you?  If it does, it would be interesting to figure out what's making a difference.

@dot-asm, can I ask you for your wisdom?  (Since you're the author of the fix I backported in #2701)

**Comment by rodwiddowson** [comment]
Not sure that this matters but I'll flag this - from reading above the issue is not in the specific VC runtime for VC15, it is in the Universal Run Time.  This is distributed via windows update.  So this at least in theory this for not "VC15 builds " but for "any system using the UCRT runtime" which I guess means >=VC15.

What spooks me is that by this not being tied to the specific runtime the barrier against changes in the ABI and the implementation has been raised to the same level as the SysCall interface (or Win32).  

So if a a change to provoke this sort of issue was introduced it would happen to end users "just because" of windows update. Then we'd not notice it until otherwise working installations suddenly broke,  This is rather different from qualifying against a specific C run time and shipping with it.

**Comment by dot-asm** [comment]
> #2701 didn't worked for me.

Define "didn't work". Do note that [as discussed in RT reply] right thing to do is to switch from _fp_ interfaces to corresponding _bio_ ones. Well, one can naturally argue that if you have working application targeting LTS[!] 1.0.2 you are entitled for solution fully compatible with the way things work currently. Or in other words that we can play fp->bio hardball only with users who are migrating to 1.1.x. With this in mind just back-porting fix from master to 1.0.2 can actually be considered insufficient.

As for suggestion from Microsoft. They are kind of in privileged position, because they have inside information and in position to affect direction of future development. Limiting to bare minimum of `std{in|out|err}` in 1.1.x is effectively recognition of the fact that we don't same opportunity. I mean we prefer [and should do so] to stick to something we can rely on, i.e. to something Microsoft is obliged to implement to comply with standards.

Bottom line is that I'd argue in favour of keeping bare minimum of `std{in|out|err}`, but it's probably appropriate to adopt Microsoft-style approach for 1.0.2. I can look into it...

**Comment by dot-asm** [comment]
Having written this I realized that there is kind of contradiction. If one argues in favour of limiting to bare minimum of `std{in|out|err}`, then why does it have to be specific to VS>=2015? I mean right thing to do is to limit it to the bare minimum for *all* Windows compilers.

**Comment by sunnyap** [comment]
@dot-asm #2701 did worked for me. I have to start fresh, re-compile and test. Thanks.

**Comment by dot-asm** [comment]
Had closer look. Unfortunately bare minimum of `std{in|out|err}` is really best one can do in VS2015 context. In other words it's impossible to achieve functional equivalence in 1.0.2, and #2701 is really the only thing one can do. So +1 for that. As for limiting to `std{in|out|err}` for *all* Windows compilers. Question is if it's appropriate to do in 1.1.1 or does it have to wait till 1.2. One can probably argue either way...

To recap, applink becomes handy when you mix compiler versions *or* flags. As for flags. For example if you have OpenSSL dlls and want to produce debug build of your application *without* having to build even debug version of OpenSSL dlls, applink allows you to do that. To debug application without having it crash that is. Inherently undesired effect in VS2015 context is that applink will get engaged when it's not actually necessary, i.e. even when both application and dlls are compiled with same flags. It might be possible to mitigate this specific problem by pulling list of dlls at run-time and seeing if there are multiple crts. Or rather if there is only one crt found in current application context, then never engage applink. Would that work? Hmm... It's impossible to detect crt copy linked statically into application... So probably no, at least it wouldn't be fool-proof...

**Comment by levitte** [comment]
Incidently, #2701 got merged to 1.0.2 earlier today

**Comment by sunnyap** [comment]
Yeah +1 for #2701 . This solution is in sync with recent changes in VS2015 which comes with Universal C Runtime. 

An application including OpenSSL build with VS2015 ONLY needs 2015 CRT and if its build with same compiler flags then there no need to engage applink. Users having this arrangement have no option to not engage applink (or do we have by removing  $cflags.=" -DOPENSSL_USE_APPLINK";)  Also applink is not needed in case an application only uses BIO functions. Although BIO functions was introduced long back, OpenSSL Application like CURL build with VS2015 continue to use File * I/O and I have seen it failed with VS2015 recent changes in _IOB_ENTRIES (3 from 20) and worked with 1. #2701 & 2. Replacing File * I/O with BIO functions. Question is with recent changes in VS2015 _IOB_ENTRIES reduced to 3 from 20 or bare minimum std{in|out|err}, do we still need to engage applink in VS2015 or later release?

**Comment by levitte** [comment]
In other words, are we good enough to close this issue?

**Comment by mattcaswell** [comment]
No further comment, so it appears this fix was good.
